### ISSUE TITLE
Inversion raises Exception:  "energy from equality constraints is NaN!"

### ISSUE THREAD

**Issue by chrisbc** [GNS-Science/nshm-nz-opensha#24]
This exception raised in inversion phase while running new [scriptCrustalInversionRunner](https://github.com/GNS-Science/nshm-nz-opensha/blob/chris-inversion-101/src/nz/cri/gns/NSHM/opensha/scripts/scriptCrustalInversionRunner.java)

The same script has previously run OK with smaller list of fault subsections. It's based on recent Demo03 but using only NZ crustal faults produced with latest NZCFM export.  

@kevinmilner I'm cc'ing you in case you know something about this type of error. 

Meanwhile I'm trying to get a narrow this down by reducing number of fault sections (down fom 675)
 
```
Exception in thread "main" java.lang.IllegalStateException: energy from equality constraints is NaN!
        at com.google.common.base.Preconditions.checkState(Preconditions.java:507)
        at scratch.UCERF3.simulatedAnnealing.SerialSimulatedAnnealing.calculateEnergy(SerialSimulatedAnnealing.java:304)
        at scratch.UCERF3.simulatedAnnealing.SerialSimulatedAnnealing.setup(SerialSimulatedAnnealing.java:147)
        at scratch.UCERF3.simulatedAnnealing.SerialSimulatedAnnealing.<init>(SerialSimulatedAnnealing.java:117)
        at scratch.UCERF3.simulatedAnnealing.ThreadedSimulatedAnnealing.<init>(ThreadedSimulatedAnnealing.java:112)
        at nz.cri.gns.NSHM.opensha.scripts.scriptCrustalInversionRunner.main(scriptCrustalInversionRunner.java:262)
crustal_inversion.pbs.e34867 
PS C:\Users\chrisch\beavan>
```

## Script configuration

```
#PBS -l nodes=1:ppn=32
#PBS -l walltime=04:30:00

export PATH=/opt/sw/java/java-se-8u41-ri/bin:$PATH
export JAVA_CLASSPATH=~/NSHM/opensha/nshm-nz-opensha/build/libs/nshm-nz-opensha-all.jar
export CLASSNAME=nz.cri.gns.NSHM.opensha.scripts.scriptCrustalInversionRunner
export OPTS="-fsdFile ./data/FaultModels/NZ_CFM_V0.3_crustal_opensha.xml -o ./tmp -maxFaultSections 1000 -maxLength 0.5 -maxDistance 0.1"
export OPTS=${OPTS}" -inversionMins 120 -syncInterval 30 -runInversion true"

cd ~/NSHM/opensha/nshm-nz-opensha
java -Xms4G -Xmx100G -classpath ${JAVA_CLASSPATH} ${CLASSNAME} ${OPTS} > ./tmp/crustal.log
```

## more info

 * Rupture outputs and stdout +  stderr
      https://drive.google.com/drive/folders/16UKa1E6ewiRpEnwe8zidL55yfy5OP0wn?usp=sharing

 * job history
      https://docs.google.com/spreadsheets/d/1pkL_5egm3wZ-7AE12qpb0N6gQKKnvS__5XGxz6fCt0I/edit?usp=sharing

**Comment by kevinmilner** [comment]
Do you have any zero or NaN slip rates?

**Comment by chrisbc** [comment]
It's certainly possible - this is a new data set.

**Comment by chrisbc** [comment]
Here's an example of 0 slip rate. there are a few like this, but I don't think this alone is the porblem as I succesfully processed this data already.... 

```
<i0 sectionId="0" sectionName="Ahuriri" aveLongTermSlipRate="0.0" slipRateStDev="0.0" aveDip="60.0" aveRake="100.0" aveUpperDepth="0.0" aveLowerDepth="12.0" aseismicSlipFactor="0.0" couplingCoeff="1.0" dipDirection="306.5" parentSectionId="-1" connector="false">
      <FaultTrace name="Ahuriri">
        <Location Latitude="-43.7945" Longitude="169.9948" Depth="0.0000"/>
        <Location Latitude="-43.8877" Longitude="169.9254" Depth="0.0000"/>
        <Location Latitude="-43.9570" Longitude="169.8191" Depth="0.0000"/>
        <Location Latitude="-44.0998" Longitude="169.6688" Depth="0.0000"/>
      </FaultTrace>
    </i0>
```

Is it possible to input NaN via the XML loader?

**Comment by chrisbc** [comment]
Steps tp reproduce the fault.

 - pull latest from branch https://github.com/GNS-Science/nshm-nz-opensha/tree/chris-inversion-101
 - run the script [scriptCrustalInversionRunner](https://github.com/GNS-Science/nshm-nz-opensha/tree/chris-inversion-101/src/main/java/nz/cri/gns/NSHM/opensha/scripts) with these parameters
    ```
    -fsdFile ./data/FaultModels/CFM_DEMO_crustal_opensha.xml
    -o /tmp
    -maxFaultSections  55
    -skipFaultSections 154
    -maxLength 0.5
    -maxDistance 0.5
    -inversionMins 1
    -syncInterval 5
    -runInversion true
    ``` 

   This will raise the NaN error

- now set `skipFaultSections 155` and rerun which should succeed, demonstrating that  something in section i154 (n-1) is at fault.
    ```
    <i154 sectionId="154" sectionName="Hikurangi Wellington" aveLongTermSlipRate="25.0" slipRateStDev="5.0" aveDip="0.0" aveRake="90.0" aveUpperDepth="0.0" aveLowerDepth="30.0" aseismicSlipFactor="0.0" couplingCoeff="1.0" dipDirection="336.4" parentSectionId="-1" connector="false">
      <FaultTrace name="Hikurangi Wellington">
        <Location Latitude="-42.1876" Longitude="175.0686" Depth="0.0000"/>
        <Location Latitude="-41.9563" Longitude="175.5805" Depth="0.0000"/>
        <Location Latitude="-41.7962" Longitude="176.2206" Depth="0.0000"/>
        <Location Latitude="-41.4191" Longitude="177.2835" Depth="0.0000"/>
        <Location Latitude="-41.1677" Longitude="177.7339" Depth="0.0000"/>
      </FaultTrace>
    </i154>
    ```

**Comment by chrisbc** [comment]
Some debugging reveals that this is a result of aveDip="0.0" in source data, which results in a FaultSection.getArea() returning **Infinite**. We'll workg=around for now but raise a pull request next on opensha-ucerf3 soon.

**Comment by kevinmilner** [comment]
@chrisbc do you really have faults with dip=0 or was that a placeholder?

**Comment by chrisbc** [comment]
Hi @kevinmilner  I've just been looking at the implications of that with Kiran @thingbaijam - it needs more discussion and investigation of our datasets. 

I think there are a few Hikuragni subsections twhere dip will be at or close to 0. 

Not urgent as we'll work around it until we understand things a bit more :)

**Comment by kevinmilner** [comment]
Oh this is that upper edge where it kind of bends over, interesting. It goes past zero, right? In other words, it starts dipping the other way? How are you handling that?

**Comment by thingbaijam** [comment]
> Oh this is that upper edge where it kind of bends over, interesting. It goes past zero, right? In other words, it starts dipping the other way? How are you handling that?

Not yet decided/examined as to how to deal with that, as our immediate focus is on crustal faults only.

**Comment by chrisbc** [comment]
> Oh this is that upper edge where it kind of bends over, interesting. It goes past zero, right? In other words, it starts dipping the other way? How are you handling that?

That's a good question actually. I just checked our CSV data for Hikurangi, and there are no 0 or negative dip values, which  explains why we haven't dealt with this issue on the earlier subduction/inversion tests. 

I guess Andy has ensured that  'top edge' of the tile (ie the one with the surface trace coords) is always that with the smallest z value. @andy22b can you check that please?

FYI for subduction data  we're using [this method](https://github.com/GNS-Science/nshm-nz-opensha/blob/4ec63a061214085f87fa8236eb095fc02e9822b4/src/main/java/nz/cri/gns/NSHM/opensha/ruptures/downDipSubSectTest/DownDipSubSectBuilder.java#L82)  to build the subsections from [/subduction_tile_parameters.csv](https://github.com/GNS-Science/nshm-nz-opensha/blob/chris-inversion-101/data/FaultModels/subduction_tile_parameters.csv). And we have an open [issue logged](https://github.com/GNS-Science/nshm-nz-opensha/issues/6#issuecomment-680423577) to tidy this code up a bit!

See [Demo03_RupturesAndInversion.java](https://github.com/GNS-Science/nshm-nz-opensha/blob/chris-inversion-101/src/main/java/nz/cri/gns/NSHM/opensha/demo/Demo03_RupturesAndInversion.java) for how we're using this.

**Comment by andy22b** [comment]
Yep, I'm pretty sure that the top of the patch will be the side with the hallowest depth... There is no constraint for (e.g.) the SE side being the top trace. Not sure how the code would deal with purely horizontal subduction tiles.

**Comment by chrisbc** [comment]
to fix in issue #34
